### DISCUSSION TITLE
Get initial props on server only on initial document request

### DISCUSSION THREAD

**Discussion by ekwoka** [vercel/next.js#45078] [Help] [answered]
We have a webapp that is mostly just an SPA, and we would like just the initial document request to handle seeing if the user is logged in so that the app can be immediately logged in without waiting for a network request on the browser side. this should apply for ALL routes.

I tried multiple options but it seems like this is just totally not a supported thing at all?

_app only always getInitialProps which will also run in the client on every navigation. This won't work since we need to access the DB. Can't even have the code branch here since it seems like webpack will still try to bundle in all the server code and fail

_doc getInitialProps won't work since we don't have access to the req object (oddly, the document context has a req key but it's empty.

How in the hell am I supposed to do this?

Is there a way to get the req in the doc context, or a way to more aggressively force webpack to prune out the server code in _app.getInitialProps (effectively making it a passthrough function)?

I've been all over trying to find solutions and it seems this is a user case that should be more likely, but seemingly isn't? Is it really just more everything from pages to `app`?

**Comment by icyJoseph** [comment, marked as answer]
> _app only always getInitialProps which will also run in the client on every navigation. This won't work since we need to access the DB. Can't even have the code branch here since it seems like webpack will still try to bundle in all the server code and fail

It is possible to tell webpack to not bundle certain things for the "browser path". You'd basically do a check for `req` and `res` presence in the `getInitialProps` callback. These are undefined when it runs on the browser.

> _doc getInitialProps won't work since we don't have access to the req object (oddly, the document context has a req key but it's empty.

No logic, or data fetching should go here.

> How in the hell am I supposed to do this?

With `_app`, or via middleware, which runs on every request. You'd basically do the same check you'd do for `_app`.

## Using _app

This is a bit of a hack, so study it carefully, it works, but it is hacky af... I'd choose going with middleware first, but if that's not allowed, well...

```ts
// rest of your MyApp, custom _app code goes above this part

/**
 * Disables any optimization (build time render)
 * - Except for pages that implement `getStaticProps`.
 */
async function getInitialProps({ctx}: AppContext) {
    /**
     * Server Side
     */
    if (ctx.req && ctx.res) {
        try {
            const {aFunction} = await import('some/module/of-yours'); // important name

            return {
                /* some props */
            };
        } catch (reason) {
            // always handle errors!
            return {
                /* other props in case of error */
            };
        }
    }
    /**
     * Client Side
     */

    // this branch might not be necessary if you don't return any props
    // but the idea is to avoid extra re-rendering by at least
    // providing similar data to that which the server sent

    // Another technique to avoid re-rendering would be to set the props
    // passed by this function into MyApp, as React state, and never update
    // that state - sort of stale by design, but good luck explaning that
    // to your co-workers

    // sanity check
    if (typeof window === 'undefined') throw new Error('Browser exclusive code run on Server');

    const element = document.getElementById('__NEXT_DATA__');

    const content = element?.textContent;

    if (!content) throw new Error('Problem Reading __NEXT_DATA__');

    try {
        const data = JSON.parse(content);
        // select from data props that match what the server would've sent
        // - if you are always returning {}, or no props at all then it is your lucky day
        // this step might not even be necessary

        return {
            /* some props */
        };
    } catch (reason) {
        /* handle failure of parsing content */
        return {
            /* some props */
        };
    }
}

MyApp.getInitialProps = getInitialProps;

export default MyApp;

```

And now in `next.config.js`, you'd add this for webpack (nextRuntime is `edge`, `node-js` or `undefined`):

```js
  webpack(config, { nextRuntime }) {
    if (typeof nextRuntime === "undefined") { // we are in a browser bundle
      // manually ignore some local modules and node_modules
      const { IgnorePlugin } = require("webpack");
      const ignoreUserModule = new IgnorePlugin({
        resourceRegExp: /some\/module\/of-yours/, // so we tell webpack to never bundle our module for the browser
      });

      config.plugins.push(ignoreUserModule);
    }

    return config;
  },
```

**Reply by ekwoka** [reply]
Thank you so much.

I know document shouldn't be used for logic, but it was in the "maybe this could work"? it's just very strange to me that this seems like a totally reasonable use case that is just totally not covered at all.
